File: leetcode/course-schedule-ii/course-schedule-ii-dfs-rec.rb
# @param {Integer} num_courses
# @param {Integer[][]} prerequisites
# @return {Integer[]}
def find_order(num_courses, prerequisites)
  Graph.new(num_courses, prerequisites).reverse_toposort || []
end

class Graph
  def initialize(vertex_count, edges)
    @adj = Array.new(vertex_count) { [] }
    edges.each { |src, dest| @adj[src] << dest }
  end

  def reverse_toposort
    out = []
    vis = [:white] * @adj.size

    dfs = ->(src) do # Returns true on success, false on failure due to cycle.
      case vis[src]
      when :white
        vis[src] = :gray
        return false unless @adj[src].all?(&dfs)
        out << src
        vis[src] = :black
        true
      when :gray
        false
      when :black
        true
      else
        raise 'Bug: unrecognized visitation state'
      end
    end

    (0...@adj.size).all?(&dfs) ? out : nil
  end
end

File: leetcode/course-schedule-ii/course-schedule-ii-dfs-rec.java
import java.util.Optional;

class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        var graph = new Graph(numCourses);
        for (var edge : prerequisites) graph.addEdge(edge[0], edge[1]);

        return graph.reverseToposort()
                .orElse(Collections.emptyList())
                .stream()
                .mapToInt(v -> v)
                .toArray();
    }
}

class Graph {
    Graph(int vertexCount) {
        _adj = new ArrayList<>(vertexCount);
        while (_adj.size() < vertexCount) _adj.add(new ArrayList<>());
    }

    void addEdge(int src, int dest) {
        _adj.get(src).add(dest);
    }

    Optional<List<Integer>> reverseToposort() {
        List<Integer> out = new ArrayList<>(order());
        var vis = freshVisitationList();

        // Returns true on success, false on failure due to cycle.
        var dfs = new IntPredicate() {
            @Override
            public boolean test(int src) {
                switch (vis[src]) {
                case WHITE:
                    vis[src] = Color.GRAY;
                    if (!_adj.get(src).stream().allMatch(this::test))
                        return false;
                    vis[src] = Color.BLACK;
                    out.add(src);
                    return true;

                case GRAY:
                    return false;

                case BLACK:
                    return true;

                default:
                    throw new AssertionError("unrecognized visitation state");
                }
            }
        };

        return vertices().allMatch(dfs) ? Optional.of(out) : Optional.empty();
    }

    private int order() {
        return _adj.size();
    }

    private IntStream vertices() {
        return IntStream.range(0, order());
    }

    private Color[] freshVisitationList() {
        return Stream.generate(() -> Color.WHITE)
                     .limit(order())
                     .toArray(Color[]::new);
    }

    private final List<List<Integer>> _adj;
}

enum Color {
    WHITE,
    GRAY,
    BLACK,
}
